File: demo-vk-rest-scraping.java/src/main/java/demo/reader/RepeatableReader.java
package demo.reader;

public class RepeatableReader<T extends RepeatableReader.Responsive, S> extends Reader<T> {

    RepeatableReader(Class<T> clazz) {
        super(clazz);
    }

    public T read(String urlBody) {
        T answer;
        do {
            answer = super.read(urlBody);
        } while (answer.response() == null);
        return answer;
    }

    public interface Responsive<S> {

        S response();
    }
}

File: demo-vk-rest-scraping.kotlin/src/main/kotlin/demo/reader/RepeatableReader.kt
package demo.reader

open class RepeatableReader<T : RepeatableReader.Responsive<*>, S> internal constructor(clazz: Class<T>) : Reader<T>(clazz) {

    override fun read(urlBody: String): T {
        var answer: T
        do {
            answer = super.read(urlBody)
        } while (answer.response() == null)
        return answer
    }

    interface Responsive<out S> {

        fun response(): S?
    }
}
